Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -519,6 +524,117 @@ extension SwiftLanguageServer {
+      req.reply(nil)
+      return
+    }
+
+    let skreq = SKRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.editor_open
+    skreq[keys.name] = "FoldingRanges:" + snapshot.document.url.path
+    skreq[keys.sourcetext] = snapshot.text
+    skreq[keys.syntactic_only] = 1
+
+    let handle = sourcekitd.send(skreq) { [weak self] result in
+      guard let self = self else { return }
+      guard let dict = result.success else {
+        req.reply(.failure(result.failure!))
+        return
+      }
+
+      guard let syntaxMap: SKResponseArray = dict[self.keys.syntaxmap],
+            let substructure: SKResponseArray = dict[self.keys.substructure] else {
+        return req.reply([])
+      }
+
+      var ranges: [FoldingRange] = []
+
+      // Merge successive comments into one big comment by adding their lengths.
+      var currentComment: (offset: Int, length: Int)? = nil
+
+      syntaxMap.forEach { _, value in
+        if let kind: sourcekitd_uid_t = value[self.keys.kind],
+           kind.isCommentKind(self.values),
+           let offset: Int = value[self.keys.offset],
+           let length: Int = value[self.keys.length]
+        {
+          if let comment = currentComment {
+            if offset == comment.offset + comment.length {
+              currentComment?.length += length
+            } else {
+              self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toArray: &ranges)
+              currentComment = (offset: offset, length: length)
+            }
+          } else {
+            currentComment = (offset: offset, length: length)
+          }
+        }
+        return true
+      }
+
+      // Add the last stored comment.
+      if let comment = currentComment {
+        self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toArray: &ranges)
+        currentComment = nil
+      }
+
+      var structureStack: [SKResponseArray] = [substructure]
+      while let substructure = structureStack.popLast() {
+        substructure.forEach { _, value in
+          if let offset: Int = value[self.keys.bodyoffset],
+             let length: Int = value[self.keys.bodylength],
+             length > 0
+          {
+            self.addFoldingRange(offset: offset, length: length, in: snapshot, toArray: &ranges)
+          }
+          if let substructure: SKResponseArray = value[self.keys.substructure] {
+            structureStack.append(substructure)
+          }
+          return true
+        }
+      }
+
+      req.reply(ranges)
+    }
+
+    // FIXME: cancellation
+    _ = handle
+  }
+
+  func addFoldingRange(offset: Int, length: Int, kind: FoldingRangeKind? = nil, in snapshot: DocumentSnapshot, toArray ranges: inout [FoldingRange]) {
+    let capabilities = clientCapabilities.textDocument?.foldingRange
+    if let rangeLimit = capabilities?.rangeLimit, ranges.count == rangeLimit {
+      return
```

Comment:
Thinking about it now, I should've halted the whole request if the limit was reached.

Reply:
That seems like a good idea to me; would be interesting to know what other language servers do.